Flask web app blog (#1664)
oktadeveloper/okta-blog: _source/_posts/2026-07-14-flask-oauth-web-app.md

---
layout: blog_post

> title: "How to Build a Secure Flask App with Okta and a Custom Resource Server"

Reviewer: this is a very strange title. are you sure the term custom resource server is required here? wouldn't it be the default resource server for most apps? i think we need to evaluate this

Author: Thanks for the feedback, @alisaduncan! 
How about this: 
Build a secure Flask app with OAuth 2.1 authentication and a protected API with Okta

Reviewer: No, definitely not that.

I feel like the title is strange because we're missing what the tutorial is about. Why are we calling a protected API? 

I think what you have here is valuable for the community, but it's just not clicking. For one, i think you're hyper-focused on PKCE when it shouldn't be such a big focal point here.

author: akanksha-bhasin
by: advocate
communities: [python]
description: "Learn to add secure authentication to a Flask app with Okta and call a protected API using scoped access tokens."
tags: [python, flask, oauth, oidc, pkce, authentication]
type: conversion
image: blog/flask-oauth-web-app/flask-oauth-web-app-social-image.jpg
github: https://github.com/oktadev/okta-flask-oauth-example
---

> Python syntax and the flexibility of the Flask microframework make it a popular choice for quickly building web applications. While Flask provides the essentials to get you started, you'll need to implement a critical piece yourself: secure user authentication. After all, how do you securely sign users into your application?

Reviewer: I felt a bit lost during this tutorial. I think the missing piece here is why do you have this random API floating out separate from your Flask app? I think setting up that use case is important. this whole first part is all about secure authentication. then making the call to a resource server is thrown in as an after thought.

> This tutorial shows you how to solve that problem. You'll add a secure authentication layer to your Flask application using modern standards like OAuth 2.1 and OpenID Connect (OIDC), following the latest security best practices. You'll also learn how to implement the secure Authorization Code flow with Proof Key for Code Exchange (PKCE) and build a complete sign-in and sign-out workflow. Second, and just as importantly, you'll learn how to use the resulting scoped access tokens to make protected calls to an API.

Reviewer: also let's be clear, you are using a library that handles the OAuth handshake securely - you aren't implementing auth code flow with PKCE, so let's update wording to reflect it. this is a good thing! it's best practice to use an oidc client library!

Reviewer: Also don't access tokens have some scope? If the goal of this post is scope, you should be talking about scope vs use privileges (or permissions). Those shouldn't be folded into scopes

Author: I want to make sure I address this correctly, but I'm not fully clear on what you mean by "scope vs use privileges (or permissions)", could you please clarify it more

Check out the complete source code on [GitHub](https://github.com/oktadev/okta-flask-oauth-example) and get started without setting it up from scratch.

**Table of Contents**{: .hide }
* Table of Contents
{% include toc.md %}

> ## Build a Flask app with OAuth 2.0 authentication and a protected API

Reviewer: Please update this title. we're not using Oauth authtneication, you usig OIDC.

In this tutorial, you'll build a simple dashboard application and learn how to:

* Securely sign users in to view their profile information using Okta as the OpenID Connect identity provider.  
* Implement the Authorization Code flow with [PKCE](https://oauth.net/2/pkce/) to adhere to the latest OAuth 2.1 security best practices.  
* Add a feature for new users to self-register for an account directly from the sign-in page.  

> * Enable authenticated users to interact with a protected API using OAuth 2.0 access tokens.

Reviewer: ```suggestion
* Enable authenticated users to interact with a protected API using access tokens.
```

> ## Prerequisites

Reviewer: We should not have a header with just the word "Prerequeisites" How about turning this into a bold?

## Add secure user login to a Flask application with Okta

Now that you've configured your app integration in Okta and noted the required Okta application details, it's time to create your Flask application.

### Create a basic Flask web application

First, you'll create a project folder and a Python virtual environment, then install the necessary dependencies.

1. Create a project folder named `flask-web-app` and a subfolder called `venv`.  
2. Open a terminal window in the project folder and create a Python virtual environment

   ```bash
   python -m venv venv
   ```

3. Activate the virtual environment. Use the command that corresponds to your operating system.

   ```bash
   # On macOS/Linux
   source venv/bin/activate

   # On Windows
   venv\Scripts\activate.bat
   ```

4. With your environment active, install the required libraries:

   ```bash
   pip install Authlib==1.6.1 Flask[async]==3.1.1 python-dotenv==1.1.1 requests==2.32.4 okta-jwt-verifier==0.3.0
   ```

### Configure Flask environment variables

> To keep sensitive credentials secure, you'll store them in a `.env` file and load them into the application's configuration. This practice prevents hard-coding secrets directly into your source code.

Reviewer: i think you need to explicitly state to add this file to gitignore.

Also, i disagree with the statement that this keeps sensitive creds secure. for that, there should be some sort of secrets manager. in this post, i think the correct language here is 
> To keep sensitive credentials out of source control...

```python
from flask import Flask, render_template, url_for, redirect, session
from authlib.integrations.flask_client import OAuth
from dotenv import load_dotenv
import os
import requests

load_dotenv()

app = Flask(__name__)
app.secret_key = os.urandom(24)

app.config.from_object('config')

@app.route("/")
def index():
    return render_template('index.html')

if __name__ == "__main__":
    app.run(debug=True)
```

#### Initialize the Okta OIDC client

To set up the connection to Okta, add the following code to the bottom of your `app.py` file. This code initializes an OAuth object and registers the Okta application as a remote provider.

```python
# Create a registry with an OAuth object
oauth = OAuth(app)

# Register a remote application
oauth.register(
    name='okta',
    client_id=app.config['OKTA_CLIENT_ID'],
    client_secret=app.config['OKTA_CLIENT_SECRET'],
    server_metadata_url=app.config['OKTA_METADATA_URL'],
    client_kwargs={
        'scope': 'openid profile email offline_access',
        'code_challenge_method': 'S256'
    }
)
```

> The first parameter in the register() method (in this case, *Okta*) is the name of the remote application. You'll later access the remote application with `oauth.okta` to handle all the OIDC-related functions.

Reviewer: ```suggestion
The first parameter in the `register()` method (in this case, *Okta*) is the name of the remote application. You'll later access the remote application with `oauth.okta` to handle all the OIDC-related functions.
```

The client\_kwargs dictionary passes extra parameters to the authorization request. Because the code\_challenge\_method is present, the Authlib library automatically uses the [PKCE](https://oauth.net/2/pkce/) flow, which enhances security.

#### Build Flask routes for OIDC authentication

The sign-in process begins when the user clicks the Login button, which redirects the browser to the Okta-hosted Sign-In page.

> After the user signs in, Okta redirects the browser to the configured sign-in redirect URI, sending user profile information to the application. Similarly, after a user signs out, Okta redirects the browser to the sign-out redirect URI.

Reviewer: "sending user profile info to the application" not quite correct. Do you mean the client now has an id token which contains a minimum set of user identity claims?

if you want to get all of the info held within the userprofile endpoint, that's a separate call.

Author: Updated the paragraph to reflect it correctly

Let's start by creating a simple HTML page that allows the user to log in. This page will display a sign-in button for signed-out users and a sign-out button for logged-in users, along with profile information.

1. Add the `/login` route to your `app.py` file. This endpoint serves as the entry point for the authentication flow. It constructs the authorization redirect URI and sends the user's browser to the Okta-hosted Sign-In page to begin the OIDC process.

   ```python
   @app.route('/login')
   def login():
       redirect_uri = url_for('authorization_code_callback', _external=True)
       return oauth.okta.authorize_redirect(redirect_uri)
   ```

> 2. Next, implement the callback handler for the Redirect URI you configured in your Okta application (`http://localhost:5000/authorization-code/callback`). This handler is crucial as it processes the authorization\_code returned by Okta after a successful sign-in. It performs the code-for-token exchange to get an access token and ID token, then stores the userinfo response alongside the tokens in the session.

Reviewer: Does this make the userinfo call? or are you getting id claims from the id token?

Also, should probably mention that the library creates a session?

Author: I am using the userinfo endpoint, and I have added the details about session creation. Thanks!

   ```python
   @app.route('/logout')
   def logout():
       token = session.get('token')

       id_token_hint = None
       if token:
           id_token_hint = token.get('id_token')

       conf = oauth.okta.server_metadata
       logout_endpoint = conf.get('end_session_endpoint')

       session.pop('loggedInUser', None)
       session.pop('token', None)

       if logout_endpoint:
           logout_url = (
               f'{logout_endpoint}?'
               f'id_token_hint={id_token_hint}&'
               f'post_logout_redirect_uri={url_for("index", _external=True)}'
           )
           return redirect(logout_url)

       return redirect('/')
   ```

4. Finally, update the root route (`/`). This route now checks the session for a signed-in user and passes that information to the template.

   ```python
   @app.route("/")
   def index():
       loggedInUser = session.get('loggedInUser')
       return render_template('index.html', user=loggedInUser)
   ```

> With the back-end routes in place, you can now create the front-end files to provide a user interface for signing in, signing out, and displaying profile data.

Reviewer: i recommend sayng "create the html template files" isntead?

  font-size: 1rem;
  text-decoration: none;
  display: inline-block;
  margin-top: 1rem;
  font-weight: 600;
}

button:hover,
a.button-link:hover {
  background-color: #155ab6;
  color: #fff;
}
```

### Run the Flask OAuth web app

You're ready to test the sign-in and sign-out functionality.

* Start the application with the Flask server.

```bash
flask run
```

* Open a new terminal window and start the resource server.

```bash
flask --app resource-server.py run --port 5001
```

* Open a browser and go to http://localhost:5000.  
* Click Login. The browser redirects you to the Okta Sign-In Widget.  
* After you sign in, Okta redirects you back to your application, and the page displays the signed-in user's details.  

{% img blog/flask-oauth-web-app/flask-app-after-login.jpg alt:"Flask web app home page showing a welcome message after successful login with a Fetch Users button" width:"800" %}{: .center-image }

* Click Logout. The browser clears your session and returns you to the home page.

You've successfully added a secure authentication flow to your Flask app. Next, you'll extend it to call a protected API using scoped tokens.

## Call a protected API with scoped tokens

> Using OAuth 2.0 for API authentication is a modern security best practice. It allows you to use access tokens to interact with APIs to fetch and manage data securely. This section walks you through how to use a scoped access token to fetch data from your own resource server.

Reviewer: ```suggestion
Using OAuth for API authorization is a modern security best practice. It allows you to use access tokens to interact with APIs to fetch and manage data securely. This section walks you through how to use a scoped access token to fetch data from your own resource server.
```

To include this scope in the tokens minted from Okta, follow these steps:

* In your Okta Admin Console, navigate to **Security** \> **API**.  
* Select your **Authorization Server** and go to the **Scopes** tab  
* Click the **Add Scope** button and enter `api:read-users` in the **Name** field.  
* Optionally, enter the **Description** and the **Display Phrase**  
* Click **Create**.

Now, when this scope is requested, Okta will include it in the tokens it issues.

Before creating the resource server, add an extra layer of security. You want your endpoint only to consume tokens intended for your server, preventing tokens issued for other services from being replayed or misused.

To do this, restrict the `aud` claim that Okta returns in the access token to your resource server. To achieve this:

* In your Okta Admin Console, navigate to **Security** \> **API**.  
* Select your **Authorization Server**. In the **Settings** tab, click **Edit**  
* Change the **Audience** field to [http://localhost:5001](http://localhost:5001) — this is where your resource server will be hosted

> ### Build a custom resource server

Reviewer: This should be reworded. we use the term "custom resource server" as it has specific meaning in oauth terminology (plus an Okta flair). Use wording from the POV of the reader. This is probably their API - so it should be something like user API
